src/neurostatslib/solver.py

```diff
@@ -0,0 +1,593 @@
+    """
+
+    allowed_optimizers = [
+        "GradientDescent",
+        "BFGS",
+        "LBFGS",
+        "ScipyMinimize",
+        "NonlinearCG",
+        "ScipyBoundedMinimize",
+        "LBFGSB",
+    ]
+
+    def __init__(
+        self,
+        solver_name: str = "GradientDescent",
+        solver_kwargs: Optional[dict] = None,
+        regularizer_strength: float = 1.0,
+    ):
+        super().__init__(solver_name, solver_kwargs=solver_kwargs)
+        self.regularizer_strength = regularizer_strength
+
+    def penalization(self, params: Tuple[jnp.ndarray, jnp.ndarray]) -> jnp.ndarray:
+        """
+        Compute the Ridge penalization for given parameters.
+
+        Parameters
+        ----------
+        params :
+            Model parameters for which to compute the penalization.
+
+        Returns
+        -------
+        float
+            The Ridge penalization value.
+        """
+        return (
+            0.5
+            * self.regularizer_strength
+            * jnp.sum(jnp.power(params[0], 2))
+            / params[1].shape[0]
+        )
+
+    def instantiate_solver(
+        self,
+        loss: Callable[
+            [Tuple[jnp.ndarray, jnp.ndarray], jnp.ndarray, jnp.ndarray], jnp.ndarray
+        ],
+    ) -> Callable[
+        [Tuple[jnp.ndarray, jnp.ndarray], jnp.ndarray, jnp.ndarray], jaxopt.OptStep
+    ]:
+        """
+        Instantiate the solver with a penalized loss function.
+
+        Parameters
+        ----------
+        loss :
+            The original loss function to be optimized.
+
+        Returns
+        -------
+        Callable
+            A function that runs the solver with the penalized loss.
+        """
+        self._check_is_callable_from_jax(loss)
+
+        def penalized_loss(params, X, y):
+            return loss(params, X, y) + self.penalization(params)
+
+        solver_kwargs = self.solver_kwargs.copy()
+        solver_kwargs["fun"] = penalized_loss
+        return self.get_runner(solver_kwargs, {})
+
+
+class ProxGradientSolver(Solver, abc.ABC):
+    """
+    Solver for optimization using the Proximal Gradient method.
+
+    This class utilizes the `jaxopt` library's Proximal Gradient optimizer. It extends
+    the base Solver class, with the added functionality of a proximal operator.
+
+    Attributes
+    ----------
+    allowed_optimizers : List[...,str]
+        A list of optimizer names that are allowed to be used with this solver.
+    mask : Optional[Union[NDArray, jnp.ndarray]]
+        An optional mask array for element-wise operations. Shape (n_groups, n_features)
+    """
+
+    allowed_optimizers = ["ProximalGradient"]
+
+    def __init__(
+        self,
+        solver_name: str,
+        solver_kwargs: Optional[dict] = None,
+        regularizer_strength: float = 1.0,
+        mask: Optional[Union[NDArray, jnp.ndarray]] = None,
+    ):
+        super().__init__(solver_name, solver_kwargs=solver_kwargs)
+        self._mask = mask
```

Comment:
```suggestion
        self.mask = mask
```
Then `_check_mask` gets run automatically here, which you want (as written, `LassoSolver` accepts a mask but never checks it)

Reply:
This actually breaks the initialization of the Lasso solver, so I needed another solutioun.

**Why it breaks:**
The lasso solver inherits from ProxGradientSolver (as group lasso does) but doesn't require the mask for grouping, so I set it to `None`. 

`None` mask will raise an exception. 

**Solution:**
I moved the mask entirely in the group lasso, the only one requiring masks. 

More long term, if we want to support pytrees, this should be done all using trees, but for a first release it should be sufficient